### ISSUE TITLE
Performance of scalar vs vectorized functions

### ISSUE THREAD

**Issue by jorisvandenbossche** [azhpushkin/testgeos#1]
@azhpushkin thanks a lot for looking into this issue, and creating this test repo!

As I mentioned on twitter, I think the main issue is: you are looking at "scalar" performance, while PyGEOS is optimized for operating element-wise (vectorized) on many geometries at once. For single objects, there will indeed be some overhead. See also https://github.com/Toblerity/Shapely/pull/983#issuecomment-708313322 where we had a discussion about this last week.  
The main reason for the overhead is because we are using the numpy C API to create ufuncs, that iterate through an ndarray of any dimension and apply the function element-wise. This works on scalars out of the box as well, but for this case, it clearly has some overhead. 

As I also argued in the Shapely PR linked to above, we should certainly check this performance and fix low-hanging fruit to optimize it, but I think getting this "scalar" performance fully optimized should probably not be our main priority. The speed of scalar functions typically only matters if you are repeating it many times. And in that case, you should try to use the vectorized ufuncs API of PyGEOS.

For example, the following code using PyGEOS:

https://github.com/azhpushkin/testgeos/blob/dff74d17a688d0ab49e5820a17b71cc45fdce0d1/testgeos/pygeos_raw.pyx#L7-L16

could be rewritten as (didn't run the code, so might have made a mistake):

```
def filter(coordinates, corner_1, corner_2): 
    source_points = points(coordinates) 
    polygon = box(*corner_1, *corner_2) 
  
    return within(source_points, polygon).sum()
```

It would be interesting to compare the performance with that snippet.

**Comment by azhpushkin** [comment]
Yep, already saw that and updated in a newer version. Funny enough, now implementation over vectorized data is on par with my DIY GEOS-C integration. Wondering if I can in any way combine those and get even more performance?

**Comment by azhpushkin** [comment]
For ease of use, here are results (same are in README now)
```
Timing average time:
* GEOS and CPP: 0.004952797750047466
* PYGEOS: 0.010572109450004063
* PYGEOS via lib: 0.009557065700028034
* PYGEOS (scalar objects) with GEOS-C: 0.009862124800019956
```

raw version is still faster, but only 2x, which is expected I believe when comparing raw C++ operations and wrappers

**Comment by jorisvandenbossche** [comment]
> Funny enough, now implementation over vectorized data is on par with my DIY GEOS-C integration.

I don't think that is necessarily surprising, because both are basically doing the same (pygeos in C using the numpy ufunc API, your example manually in cython, but both are, in the end, a for loop over the geometries at the C level calling the same GEOSWithin function).

> Raw C++ version is still faster, but only 2x, which is expected I believe when comparing pure machine-code operations and wrappers

Yes, I suppose so. I don't know what exact overhead can be expected, but there for sure will always be overhead from using the C API instead of the C++ API (and improving this is probably something to be done in the GEOS project itself). 
Note that `within` is generally a quite fast function, and for more expensive operations (where more time is spent in the actual computation in GEOS, and less in the function calling machinery) I expect this overhead to be smaller as 2x

**Comment by azhpushkin** [comment]
> for more expensive operations ... I expect this overhead to be smaller as 2x
Didn't think of it from such an angle, makes even more sense now, thanks!
